src/lib/hooks/useInteractions/index.svelte.ts

```diff
@@ -0,0 +1,114 @@
+import type { HTMLAttributes } from 'svelte/elements';
+
+const ACTIVE_KEY = 'active';
+const SELECTED_KEY = 'selected';
+
+interface ExtendedUserProps {
+	[ACTIVE_KEY]?: boolean;
+	[SELECTED_KEY]?: boolean;
+}
+
+interface ElementProps {
+	reference?: HTMLAttributes<Element>;
+	floating?: HTMLAttributes<HTMLElement>;
+	item?: HTMLAttributes<HTMLElement> | ((props: ExtendedUserProps) => HTMLAttributes<HTMLElement>);
+}
+
+function mergeProps<Key extends keyof ElementProps>(
+	userProps: (HTMLAttributes<Element> & ExtendedUserProps) | undefined,
+	propsList: Array<ElementProps | void>,
+	elementKey: Key
+): Record<string, unknown> {
+	const map = new Map<string, Array<(...args: unknown[]) => void>>();
+	const isItem = elementKey === 'item';
+
+	let domUserProps = userProps;
+	if (isItem && userProps) {
+		const { [ACTIVE_KEY]: _, [SELECTED_KEY]: __, ...validProps } = userProps;
+		domUserProps = validProps;
+	}
+
+	return {
+		...(elementKey === 'floating' && { tabIndex: -1 }),
+		...domUserProps,
+		...propsList
+			.map((value) => {
+				const propsOrGetProps = value ? value[elementKey] : null;
+				if (typeof propsOrGetProps === 'function') {
+					return userProps ? propsOrGetProps(userProps) : null;
+				}
+				return propsOrGetProps;
+			})
+			.concat(userProps)
+			.reduce((acc: Record<string, unknown>, props) => {
+				if (!props) {
+					return acc;
+				}
+
+				Object.entries(props).forEach(([key, value]) => {
+					if (isItem && [ACTIVE_KEY, SELECTED_KEY].includes(key)) {
+						return;
+					}
+
+					if (key.indexOf('on') === 0) {
+						if (!map.has(key)) {
+							map.set(key, []);
+						}
+
+						if (typeof value === 'function') {
+							map.get(key)?.push(value);
+
+							acc[key] = (...args: unknown[]) => {
+								return map
+									.get(key)
+									?.map((fn) => fn(...args))
+									.find((val) => val !== undefined);
+							};
+						}
+					} else {
+						acc[key] = value;
+					}
+				});
+
+				return acc;
+			}, {})
+	};
+}
+
+interface UseInteractionsReturn {
```

Comment:
Not a fan of types or interfaces being intermixed with other contents on the page. I'd suggest moving the interface up to to the top with the other two. You can always hit `command/control + left click` on the reference to jump to the definition.

Reply:
Good catch, I agree here.